README.md
# Responsive Captcha

[![Packagist Version](https://img.shields.io/packagist/v/theodorejb/responsive-captcha.svg)](https://packagist.org/packages/theodorejb/responsive-captcha) [![License](https://img.shields.io/packagist/l/theodorejb/responsive-captcha.svg)](https://packagist.org/packages/theodorejb/responsive-captcha) [![Build Status](https://travis-ci.org/theodorejb/Responsive-Captcha.svg?branch=master)](https://travis-ci.org/theodorejb/Responsive-Captcha)

Prevent form spam by generating random, accessible arithmetic and logic questions.

Examples:

* "What is the fourth letter in snowboard?"
* "What is the sum of four and six?"
* "What is eight multiplied by two?"
* "Which is smallest: sixty-six, one hundred, or twenty-two?"

Users can respond with either the numeric or textual version of an answer (e.g. "16" or "sixteen").

For background info on this project, see my blog post: https://theodorejb.me/2012/12/30/responsive-captcha/

## Install via Composer

`composer require theodorejb/responsive-captcha`

## Usage

1. Generate a random question:

    ```php
    use function theodorejb\ResponsiveCaptcha\{randomQuestion, checkAnswer};

    $qa = randomQuestion();
    $realAnswer = $qa->getAnswer(); // save somewhere (e.g. in session or encrypted single-use token)
    ```

2. Display question in form:

    ```html+php
    <label>
	    <?= $qa->getQuestion() ?>
        <input type="text" name="captcha" />
	</label>
	```

3. Check whether the user's response is correct:

    ```php
    $answer = filter_input(INPUT_POST, "captcha");

    if ($answer !== null) {
        if (checkAnswer($answer, $realAnswer)) {
            // code to execute if the captcha answer is correct
        } else {
            // the answer is incorrect - show an error to the user
        }
    }
    ```


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [3.0.1] - 2020-08-31
### Changed
- Updated word list
- PHP 7.1+ is now required
- Optimized algorithm for getting unique random integers

## [3.0.0] - 2017-04-16
### Changed
- No longer depends on PHP sessions - you can now save answers any way you'd like
  (e.g. in single-use tokens which can be used with an API)
- PHP 7+ is now required
- Simplified API (see readme for details)
- Greatly expanded test coverage

## [2.0.2] - 2016-02-26
### Fixed
- Random numbers are now properly shuffled

## [2.0.1] - 2015-01-22
### Fixed
- Bug which prevented the textual version of a number from being accepted in an answer.

## [2.0.0] - 2014-09-21
### Added
- `getAnswer()` method

### Changed
- Moved to `theodorejb` namespace
- Install using Composer
- `checkAnswer()` now returns `true` or `false`, rather than throwing an exception

## [1.0.1] - 2013-08-01
- Initial semver release

[Unreleased]: https://github.com/theodorejb/Responsive-Captcha/compare/v3.0.1...HEAD
[3.0.1]: https://github.com/theodorejb/Responsive-Captcha/compare/v3.0.0...v3.0.1
[3.0.0]: https://github.com/theodorejb/Responsive-Captcha/compare/v2.0.2...v3.0.0
[2.0.2]: https://github.com/theodorejb/Responsive-Captcha/compare/v2.0.1...v2.0.2
[2.0.1]: https://github.com/theodorejb/Responsive-Captcha/compare/v2.0.0...v2.0.1
[2.0.0]: https://github.com/theodorejb/Responsive-Captcha/compare/v1.0.1...v2.0.0
[1.0.1]: https://github.com/theodorejb/Responsive-Captcha/tree/v1.0.1


test/ResponsiveCaptchaTest.php
<?php

declare(strict_types=1);

namespace theodorejb\ResponsiveCaptcha;

use PHPUnit\Framework\TestCase;

class ResponsiveCaptchaTest extends TestCase
{
    public function testCheckAnswer()
    {
        $this->assertTrue(checkAnswer('Five', 5));
        $this->assertTrue(checkAnswer('Five', 'five'));
        $this->assertFalse(checkAnswer('Five ', 5));
        $this->assertFalse(checkAnswer('Five', 'six'));
    }

    public function testGetWordFromNumber()
    {
        // it should work for integers between 0 and 20
        $this->assertSame("zero", getWordFromNumber(0));
        $this->assertSame("twenty", getWordFromNumber(20));

        // it should work for 100
        $this->assertSame("one hundred", getWordFromNumber(100));

        // it should work for integers between 21 and 99
        $this->assertSame("fifty-six", getWordFromNumber(56));
        $this->assertSame("eighty", getWordFromNumber(80));
    }

    public function testGetUniqueIntegers()
    {
        $array = getUniqueIntegers(5);

        // the array should contain 5 unique integers
        $this->assertCount(5, $array);
        $this->assertSame(array_unique($array), $array);
        $this->assertIsInt($array[0]);
    }

    public function testGetAdditionProblem()
    {
        $qa = getAdditionProblem(2, 2, 0);
        $this->assertSame("What is two plus two?", $qa->getQuestion());
        $this->assertSame(4, $qa->getAnswer());

        $qa2 = getAdditionProblem(7, 3, 1);
        $this->assertSame("What is the sum of seven and three?", $qa2->getQuestion());
        $this->assertSame(10, $qa2->getAnswer());
    }

    public function testGetSubtractionProblem()
    {
        $expected = "What is ten minus five?";
        $qa = getSubtractionProblem(5, 10);
        $qa2 = getSubtractionProblem(10, 5);

        // parameter order shouldn't matter
        $this->assertSame($expected, $qa->getQuestion());
        $this->assertSame($expected, $qa2->getQuestion());
        $this->assertSame(5, $qa->getAnswer());
        $this->assertSame(5, $qa2->getAnswer());
    }

    public function testGetMultiplicationProblem()
    {
        $qa = getMultiplicationProblem(4, 5, 0);
        $this->assertSame("What is four times five?", $qa->getQuestion());
        $this->assertSame(20, $qa->getAnswer());

        $qa2 = getMultiplicationProblem(0, 6, 1);
        $this->assertSame("What is zero multiplied by six?", $qa2->getQuestion());
        $this->assertSame(0, $qa2->getAnswer());
    }

    public function testGetDivisionProblem()
    {
        $qa = getDivisionProblem(7, 4);
        $this->assertSame("What is twenty-eight divided by four?", $qa->getQuestion());
        $this->assertSame(7, $qa->getAnswer());
    }

    public function testGetLetterProblem()
    {
        $qa = getLetterProblem("math", 4);
        $expected = "What is the last letter in math?";
        $this->assertSame($expected, $qa->getQuestion());
        $this->assertSame('h', $qa->getAnswer());

        $qa2 = getLetterProblem("math", 5);
        $this->assertSame($expected, $qa2->getQuestion());
        $this->assertSame('h', $qa2->getAnswer());

        $qa3 = getLetterProblem("math", 2);
        $this->assertSame("What is the third letter in math?", $qa3->getQuestion());
        $this->assertSame('t', $qa3->getAnswer());
    }

    public function testGetNumberProblem()
    {
        $qa = getNumberProblem([2, 7, 1], 0);
        $this->assertSame("Which is largest: two, seven, or one?", $qa->getQuestion());
        $this->assertSame(7, $qa->getAnswer());

        $qa2 = getNumberProblem([4, 3, 2], 1);
        $this->assertSame("Which is smallest: four, three, or two?", $qa2->getQuestion());
        $this->assertSame(2, $qa2->getAnswer());
    }
}


src/ResponsiveCaptcha.php
<?php

declare(strict_types=1);

namespace theodorejb\ResponsiveCaptcha;

/**
 * Generate a random question string and store the answer in the session
 */
function randomQuestion(): QuestionAnswer
{
    $function = random_int(0, 3);

    if ($function === 0) {
        $words = [
            "airplane",
            "basketball",
            "butterfly",
            "chocolate",
            "delicious",
            "dumpling",
            "earth",
            "elephant",
            "football",
            "generous",
            "happy",
            "helicopter",
            "island",
            "jupiter",
            "kitten",
            "laugh",
            "mirror",
            "moon",
            "music",
            "neptune",
            "orange",
            "pencil",
            "physics",
            "piano",
            "quarter",
            "railroad",
            "rainbow",
            "rocket",
            "saturn",
            "science",
            "skyscraper",
            "starfish",
            "sunshine",
            "transparent",
            "ultraviolet",
            "velocity",
            "windshield",
            "xylophone",
            "yesterday",
            "youth",
            "zebra",
        ];

        $randomWordPosition = array_rand($words);
        return getLetterProblem($words[$randomWordPosition], random_int(0, 5));
    }

    if ($function === 1) {
        return getNumberProblem(getUniqueIntegers(3), random_int(0, 1));
    }

    // get a random arithmetic question
    // get a random number between 0 and 3 to determine whether to add, subtract, multiply, or divide
    $function = random_int(0, 3);

    if ($function === 0) {
        return getAdditionProblem(random_int(0, 10), random_int(0, 10), random_int(0, 1));
    }

    if ($function === 1) {
        return getSubtractionProblem(random_int(0, 10), random_int(0, 10));
    }

    if ($function === 2) {
        return getMultiplicationProblem(random_int(0, 10), random_int(0, 10), random_int(0, 1));
    }

    return getDivisionProblem(random_int(1, 10), random_int(1, 5));
}

/**
 * Returns true if the submitted answer matches the real answer
 * @param int|string $realAnswer
 */
function checkAnswer(string $submittedAnswer, $realAnswer): bool
{
    // convert the answer to lower case and trim any whitespace
    $answer = strtolower($submittedAnswer);

    if (is_int($realAnswer)) {
        // both numeric and textual answers are acceptable
        return $answer === getWordFromNumber($realAnswer) ||
            filter_var($answer, FILTER_VALIDATE_INT) === $realAnswer;
    }

    return $answer === $realAnswer;
}

/**
 * Returns the name of any positive integer less than or equal to 100
 */
function getWordFromNumber(int $number): string
{
    if ($number < 0 || $number > 100) {
        throw new \UnexpectedValueException("Number must be between 0 and 100");
    }

    $numberNames = [
        0   => "zero",
        1   => "one",
        2   => "two",
        3   => "three",
        4   => "four",
        5   => "five",
        6   => "six",
        7   => "seven",
        8   => "eight",
        9   => "nine",
        10  => "ten",
        11  => "eleven",
        12  => "twelve",
        13  => "thirteen",
        14  => "fourteen",
        15  => "fifteen",
        16  => "sixteen",
        17  => "seventeen",
        18  => "eighteen",
        19  => "nineteen",
        20  => "twenty",
        30  => "thirty",
        40  => "forty",
        50  => "fifty",
        60  => "sixty",
        70  => "seventy",
        80  => "eighty",
        90  => "ninety",
        100 => "one hundred"
    ];

    if ($number <= 20 || $number === 100) {
        return $numberNames[$number];
    }

    // split the number into an array of digits
    [$tensPlace, $onesPlace] = str_split((string)$number);

    // get the name of the tens place
    $numberName = $numberNames[(int) "{$tensPlace}0"];

    // add the name of the ones place if it isn't zero
    if ($onesPlace !== '0') {
        $numberName .= '-' . $numberNames[(int) $onesPlace];
    }

    return $numberName;
}

/**
 * Returns a random addition problem after adding the answer to the session
 * Example: "What is the sum of five and six?"
 * @internal
 */
function getAdditionProblem(int $num1, int $num2, int $format): QuestionAnswer
{
    $num1Name = getWordFromNumber($num1);
    $num2Name = getWordFromNumber($num2);

    if ($format) {
        $q = "What is the sum of $num1Name and $num2Name?";
    } else {
        $q = "What is $num1Name plus $num2Name?";
    }

    return new QuestionAnswer($q, $num1 + $num2);
}

/**
 * Returns a random subtraction problem
 * Example: "What is eight minus four?"
 * @internal
 */
function getSubtractionProblem(int $num1, int $num2): QuestionAnswer
{
    // the smaller (or equal) number should be subtracted from the larger number
    $numbers = [$num1, $num2];
    sort($numbers); // the first array element is smaller (or equal)
    [$smaller, $larger] = $numbers;
    $smallerName = getWordFromNumber($smaller);
    $largerName = getWordFromNumber($larger);

    return new QuestionAnswer("What is $largerName minus $smallerName?", $larger - $smaller);
}

/**
 * Returns a random multiplication problem
 * Example: "What is two multiplied by seven?"
 * @internal
 */
function getMultiplicationProblem(int $num1, int $num2, int $format): QuestionAnswer
{
    $num1Name = getWordFromNumber($num1);
    $num2Name = getWordFromNumber($num2);

    if ($format) {
        $q = "What is $num1Name multiplied by $num2Name?";
    } else {
        $q = "What is $num1Name times $num2Name?";
    }

    return new QuestionAnswer($q, $num1 * $num2);
}

/**
 * Returns a random division problem
 * Example: "What is twenty divided by two?"
 * @internal
 */
function getDivisionProblem(int $quotient, int $divisor): QuestionAnswer
{
    $dividendName = getWordFromNumber($quotient * $divisor);
    $divisorName = getWordFromNumber($divisor);

    return new QuestionAnswer("What is $dividendName divided by $divisorName?", $quotient);
}

/**
 * Get a random letter position question
 * Example: "What is the fifth letter in Tokyo?"
 * @internal
 */
function getLetterProblem(string $randomWord, int $randLetterPos): QuestionAnswer
{
    $letterArray = str_split($randomWord);

    // there should be a chance of getting the last letter
    if ($randLetterPos === 5 || strlen($randomWord) <= $randLetterPos) {
        $letterPosName = 'last';
        $randLetter = end($letterArray); // get the last letter in the word
    } else {
        // ask for one of the first five letters (to keep it simple)
        $numberNames = ["first", "second", "third", "fourth", "fifth"];
        $letterPosName = $numberNames[$randLetterPos];
        $randLetter = $letterArray[$randLetterPos]; // this is the answer
    }

    return new QuestionAnswer("What is the $letterPosName letter in $randomWord?", $randLetter);
}

/**
 * For a range of three unique numbers, ask which one is largest or smallest
 * Example: "Which is largest: twenty-one, sixteen, or eighty-four?"
 * @param int[] $numbers
 * @internal
 */
function getNumberProblem(array $numbers, int $format): QuestionAnswer
{
    // make a string containing the names of the numbers (e.g. "one, two, or three")
    $numberString = '';
    $totalNumbers = count($numbers);

    for ($i = 0; $i < $totalNumbers; $i++) {
        $numberName = getWordFromNumber($numbers[$i]);
        $numberString .= ($i === $totalNumbers - 1) ? "or $numberName" : "$numberName, ";
    }

    if ($format) {
        // ask which is smallest
        sort($numbers); // so the first element contains the smallest number
        $q = "Which is smallest: $numberString?";
    } else {
        // ask which is largest
        rsort($numbers); // so the first element contains the largest number
        $q = "Which is largest: $numberString?";
    }

    return new QuestionAnswer($q, $numbers[0]);
}

/**
 * Returns an array of unique, random integers
 * @return int[]
 * @internal
 */
function getUniqueIntegers(int $howMany, int $min = 0, int $max = 100): array
{
    $numbers = [];

    while (count($numbers) < $howMany) {
        $numbers[random_int($min, $max)] = true;
    }

    return array_keys($numbers);
}


TextCaptcha.php
<?php

/**
 * Generate random text-based CAPTCHAs with simple arithmetic and logic quesions
 *
 * @author Theodore Brown
 * @version 2012.12.30
 */
class TextCaptcha {

    private $sessionVariableName = "TextCaptchaAnswer";

    public function __construct() {
        if (session_id() == '') {
            // no session has been started; try starting it
            if (!session_start())
                throw new Exception("Unable to start session");
            else
                session_regenerate_id (TRUE);
        }
    }

    /**
     * Checks whether a user's response matches the stored answer
     * 
     * @param string $answer The user's submitted response
     * @return boolean TRUE if the answer is correct
     * @throws Exception
     */
    public function checkAnswer($answer) {
        // convert the answer to lower case and trim any whitespace
        $answer = strtolower(trim($answer));

        // ensure that the session answer variable is set
        if (!isset($_SESSION[$this->sessionVariableName]))
            throw new Exception("The captcha answer session variable is not set");
        else {
            $storedAnswer = $_SESSION[$this->sessionVariableName];
            unset($_SESSION[$this->sessionVariableName]);

            // both numeric and textual answers are acceptable
            if ($answer == $storedAnswer || $answer === $this->getWordFromNumber($storedAnswer))
                return TRUE;
            else
                throw new Exception("Incorrect captcha response");
        }
        return FALSE;
    }

    /**
     * Generate a random question string and store the answer in the session
     * Important: call this method AFTER checking the user's response since it will replace the session answer variable
     */
    public function getNewQuestion() {
        $function = rand(0, 2);

        if ($function == 0)
            return $this->getLetterProblem();
        elseif ($function == 1)
            return $this->getNumberProblem();
        else {
            // get a random arithmetic question
            // get a random number between 1 and 4 to determine whether to add, subtract, multiply, or divide
            $function = rand(1, 4);

            if ($function == 1)
                return $this->getAdditionProblem(); // add
            elseif ($function == 2)
                return $this->getSubtractionProblem(); // subtract
            elseif ($function == 3)
                return $this->getMultiplicationProblem(); // multiply
            else
                return $this->getDivisionProblem(); // divide
        }
    }

    /**
     * Returns a random addition problem after adding the answer to the session
     * Example: "What is the sum of five and six?"
     * 
     * @return string A random addition problem
     */
    private function getAdditionProblem() {
        $num1 = rand(0, 10);
        $num2 = rand(0, 10);

        $this->storeAnswer($num1 + $num2);

        $num1Name = $this->getWordFromNumber($num1);
        $num2Name = $this->getWordFromNumber($num2);

        if (rand(0, 1))
            return "What is the sum of $num1Name and $num2Name?";
        else
            return "What is $num1Name plus $num2Name?";
    }

    /**
     * Returns a random subtraction problem
     * Example: "What is eight minus four?"
     * 
     * @return string A random subtraction problem
     */
    private function getSubtractionProblem() {
        // the smaller (or equal) number should be subtracted from the larger number
        $numbers[] = rand(0, 10);
        $numbers[] = rand(0, 10);
        sort($numbers, SORT_NUMERIC); // the first array element is smaller (or equal)

        $smallerNumber = $numbers[0];
        $largerNumber = $numbers[1];

        $smallerNumberName = $this->getWordFromNumber($smallerNumber);
        $largerNumberName = $this->getWordFromNumber($largerNumber);
        $this->storeAnswer($largerNumber - $smallerNumber);

        return "What is $largerNumberName minus $smallerNumberName?";
    }

    /**
     * Returns a random multiplication problem
     * Example: "What is two multiplied by seven?"
     * 
     * @return string A random multiplication problem
     */
    private function getMultiplicationProblem() {
        $num1 = rand(0, 10);
        $num2 = rand(0, 10);

        $this->storeAnswer($num1 * $num2);

        $num1Name = $this->getWordFromNumber($num1);
        $num2Name = $this->getWordFromNumber($num2);

        if (rand(0, 1))
            return "What is $num1Name multiplied by $num2Name?";
        else
            return "What is $num1Name times $num2Name?";
    }

    /**
     * Returns a random division problem
     * Example: "What is twenty divided by two?"
     * 
     * @return string A random division question
     */
    private function getDivisionProblem() {
        $quotient = rand(1, 10); // this will be the answer
        $divisor = rand(1, 5); // keep it simple
        $dividend = $quotient * $divisor;

        $dividendName = $this->getWordFromNumber($dividend);
        $divisorName = $this->getWordFromNumber($divisor);
        $this->storeAnswer($quotient);
        return "What is $dividendName divided by $divisorName?";
    }

    /**
     * Get a random letter position question
     * Example: "What is the fifth letter in Tokyo?"
     */
    private function getLetterProblem() {
        $words = array(
            "airplane",
            "basketball",
            "butterfly",
            "chocolate",
            "donkey",
            "dumpling",
            "elephant",
            "football",
            "grandfather",
            "helicopter",
            "island",
            "juniper",
            "kitten",
            "laughter",
            "mirror",
            "nation",
            "orange",
            "piano",
            "pencil",
            "quartet",
            "rainbow",
            "racecar",
            "railroad",
            "snowboard",
            "skyscraper",
            "sunshine",
            "starfish",
            "transparent",
            "ultraviolet",
            "velocity",
            "windshield",
            "xylophone",
            "yesterday",
            "yellow",
            "zebra"
        );

        $numberNames = array(
            "first",
            "second",
            "third",
            "fourth",
            "fifth"
        );

        $randomWordPosition = array_rand($words);
        $randomWord = $words[$randomWordPosition];
        $randomWordLength = strlen($randomWord);
        $letterArray = str_split($randomWord);

        // there should be a chance of getting the last letter
        if (rand(1, $randomWordLength) == $randomWordLength) {
            $letterPosName = 'last';
            $randLetter = end($letterArray); // get the last letter in the word
        } else {
            // ask for one of the first five letters (to keep it simple)
            if ($randomWordLength > 5)
                $max = 5;
            else
                $max = $randomWordLength;

            $randLetterPosition = rand(0, $max - 1);
            $randLetter = $letterArray[$randLetterPosition]; // this is the answer
            $letterPosName = $numberNames[$randLetterPosition];
        }

        $this->storeAnswer($randLetter);
        return "What is the $letterPosName letter in $randomWord?";
    }

    /**
     * For a range of three unique numbers, ask which one is largest or smallest
     * Example: "Which is largest: twenty-one, sixteen, or eighty-four?"
     */
    private function getNumberProblem() {
        $numbers = $this->getUniqueIntegers(3);

        // make a string containing the names of the numbers (e.g. "one, two, or three")
        $numberString = '';
        for ($i = 0; $i < count($numbers); $i++) {
            $numberName = $this->getWordFromNumber($numbers[$i]);
            if ($i == count($numbers) - 1) {
                // the last number
                $numberString .= "or $numberName";
            }
            else
                $numberString .= "$numberName, ";
        }

        if (rand(0, 1)) {
            // ask which is smallest
            sort($numbers); // so the first element contains the smallest number
            $this->storeAnswer($numbers[0]);
            return "Which is smallest: $numberString?";
        } else {
            // ask which is largest
            rsort($numbers); // so the first element contains the largest number
            $this->storeAnswer($numbers[0]);
            return "Which is largest: $numberString?";
        }
    }

    /**
     * Returns the name of any integer less than or equal to 100
     * 
     * @param integer $number A number no greater than 100
     * @return string The name of the integer
     */
    private function getWordFromNumber($number) {
        $numberNames = array(
            0 => "zero",
            1 => "one",
            2 => "two",
            3 => "three",
            4 => "four",
            5 => "five",
            6 => "six",
            7 => "seven",
            8 => "eight",
            9 => "nine",
            10 => "ten",
            11 => "eleven",
            12 => "twelve",
            13 => "thirteen",
            14 => "fourteen",
            15 => "fifteen",
            16 => "sixteen",
            17 => "seventeen",
            18 => "eighteen",
            19 => "nineteen",
            20 => "twenty",
            30 => "thirty",
            40 => "forty",
            50 => "fifty",
            60 => "sixty",
            70 => "seventy",
            80 => "eighty",
            90 => "ninety",
            100 => "one hundred"
        );

        // make sure the number is an integer
        if (is_int($number)) {
            if (($number >= 0 && $number <= 20) || $number == 100)
                return $numberNames[$number];
            elseif ($number < 100) {
                // split the number into an array of digits
                $numArray = array_reverse(str_split($number));
                $onesPlace = $numArray[0];
                $tensPlace = $numArray[1];

                // get the name of the tens place
                $numGroup = (int) $tensPlace . 0;
                $numberName = $numberNames[$numGroup];

                // add the name of the ones place if it isn't zero
                if ($onesPlace != 0)
                    $numberName .= '-' . $numberNames[$onesPlace];
                return $numberName;
            } else {
                throw new Exception("Number is out of range!");
            }
        }
        return FALSE;
    }

    /**
     * Store an answer in the session
     */
    private function storeAnswer($answer) {
        $_SESSION[$this->sessionVariableName] = $answer;
    }

    /**
     * Returns an array of unique integers between 0 and 100
     * 
     * @param int $howMany
     */
    private function getUniqueIntegers($howMany) {
        $min = 0;
        $max = 100;

        // ensure that the requested number of unique integers does not exceed those in the range
        if ($howMany < $max) {
            $numbers = array();
            for ($i = 0; $i < $howMany; $i++) {
                $newNum = rand($min, $max);
                while (in_array($newNum, $numbers)) {
                    $newNum = rand($min, $max);
                }
                $numbers[] = $newNum;
            }
            return $numbers;
        }
        else
            throw new Exception("Requested numbers are out of range!");
    }

}

?>